### ISSUE TITLE
Prevent guessable user accounts / user enumeration

### ISSUE THREAD

**Issue by MaxSchilling** [nhost/hasura-auth#294]
For data safety reasons, an (at least) an option to prevent user enumeration / guessable user accounts would be highly appreciated. This point was shared by our pen-testing company in a preparation discussion, that this is a common finding. 
I know that in some parts of the world the current behaviour is also desired, our field unfortunately provides some additional struggles.

Some of the current error messages in the authentication part allow it for attackers to try if a user is registered with a specific e-mail. Using / providing options to use neutral error messages.
OWASP describes it as follows:
https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/03-Identity_Management_Testing/04-Testing_for_Account_Enumeration_and_Guessable_User_Account.html

Relevant processes would be (from my current point of view):
- Login
- Signup (email/password) (e.g. the EmailAlreadyInUseError)
- Password reset

An additional thing would be to add the error code to the server errors, to handle the incoming errors by code and not by text-string (which might change). Stumbled across this while migrating from v1 to v2, where those errors texts changed for example. 

Would highly appreciate a chance to toggle / modify the error messages from the backend.

Many thanks!

**Comment by Pascal-Orthopy** [comment]
The endpoint "/user/email/send-verification-email" for resending the verification mail, is also a entry for guesable user accounts. 

It returns the following states:
- When user is not registered: "No user found"
- When user email is verified: "User's email is already verified"
- When user email is not verified: "Ok"

**Comment by MaxSchilling** [comment]
Hi, so you'd create a config variable which either returns error messages as today, or a neutral version? Sounds fine to me!

So for the neutral version, I guess it would be alright to just return success/ok messages, as long as no real technical error exists(too many requests, account locked, bad request,..). So if that's what you're thinking - I'd be happy :)

**Comment by plmercereau** [comment]
@elitan @MaxSchilling @Pascal-Orthopy please have a look at the above PR.
It introduces a `AUTH_CONCEAL_ERRORS` environment variable. When set, all the error flagged as sensitive will be replaced by a standard "invalid request" error

**Comment by MaxSchilling** [comment]
Hi @plmercereau looks great - not really good (even at reading) typescript, but from what I understand the handling in errors.ts covers basically all cases, also when e.g. a "user-not-found" error is sent by e.g. the passwort-reset.ts:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/65183496/204871185-9a5706d4-4956-4510-87c1-723aee54b5c9.png">

If so - awesome solution. Really appreciate the help! :)

**Comment by elitan** [comment]
@MaxSchilling I'm not 100% following what your questions are. Can you maybe rephrase them in a new message?

I'll try to clarify one thing:

On @plmercereau's PR, if `AUTH_CONCEAL_ERRORS=true`, Hasura Auth will respond with "The request payload is incorrect" instead of things like:

- User's email is already verified
- Email already in use
- User is disabled

**Comment by MaxSchilling** [comment]
Sounds good, thanks! Didnt see the alternative request answer - but then it should be easy to handle on client side!

**Comment by elitan** [comment]
I'm reopening this after a discussion with @plmercereau. 

I think we need to go over all endpoints and not only conceal error messages but also change the response type in some scenarios. As an example, on password reset we should always respond with `200` no matter if the email exists in the database or not, so we're not leaking information.

**Comment by MaxSchilling** [comment]
Makes sense! Makes it easier to identify actual errors then.
